packages/core/schematics/ng-generate/control-flow-migration/util.ts

```diff
@@ -515,20 +517,39 @@ export function getMainBlock(etm: ElementToMigrate, tmpl: string, offset: number
   return {start, middle, end};
 }
 
-const selfClosingList = 'input|br|img|base|wbr|area|col|embed|hr|link|meta|param|source|track';
+function generateI18nMarkers(tmpl: string): string {
+  let parsed = parseTemplate(tmpl);
+  if (parsed !== null) {
+    const visitor = new i18nCollector();
+    visitAll(visitor, parsed.rootNodes);
 
-function isInMigratedBlock(line: string, isCurrentlyInMigratedBlock: boolean) {
-  return line.match(startMarkerRegex) &&
-      (!line.match(endMarkerRegex) ||
-       line.lastIndexOf(startMarker) > line.lastIndexOf(endMarker)) ||
-      (isCurrentlyInMigratedBlock && !line.match(endMarkerRegex));
+    for (const [ix, el] of visitor.elements.entries()) {
+      // we only care about elements with children and i18n tags
+      // elements without children have nothing to translate
+      const offset =
+          ix * 2;  // accounts for the addition of the 2 marker characters with each loop.
+      if (el.children.length > 0) {
+        tmpl = tmpl.slice(0, el.children[0].sourceSpan.start.offset + offset) + startI18nMarker +
+            tmpl.slice(
+                el.children[0].sourceSpan.start.offset + offset,
+                el.children[el.children.length - 1].sourceSpan.end.offset + offset) +
+            endI18nMarker +
+            tmpl.slice(el.children[el.children.length - 1].sourceSpan.end.offset + offset);
+      }
```

Comment:
nit: for readability we can extract some values into consts:
```suggestion
        const startPos = el.children[0].sourceSpan.start.offset + offset;
        const endPos = el.children[el.children.length - 1].sourceSpan.end.offset + offset;
        tmpl = tmpl.slice(0, startPos) + startI18nMarker +
            tmpl.slice(startPos, endPos) +
            endI18nMarker +
            tmpl.slice(endPos));
      }
```

I'm also wondering if we can move this logic to a separate function (like `addI18nMarkers`).